core/src/main/java/org/infinispan/stream/impl/local/AbstractLocalCacheStream.java

```diff
@@ -294,9 +295,16 @@ public boolean noneMatch(Predicate<? super R> predicate) {
 
    @Override
    public CloseableIterator<R> iterator() {
-      return Closeables.iterator(getOrCreateStream());
+      // If the stream is null we can assume no intermediate operations and thus we can support remove
+      if (stream == null) {
+         return removableIterator(Closeables.iterator(getOrCreateStream()));
```

Comment:
So you are proposing a new createStream method.  That should work, not as pretty but then we wont' have another null check.